_posts/2021-05-27-TC-Guard.md

```diff
@@ -0,0 +1,162 @@
+It also allowed us to work with scrum using the built-in boards, which ensured that the mentors could track our progress.
+Overall, Azure DevOps is a very useful and feature-rich platform to use.
+
+# API
+Our first experience with Java was in highschool, where we learned most of the programming basics. 
+In college we only used Java during our Android Development course.
+Because of this, we weren't that well versed when it came down to Java specific features.
+While working on the API we learned quite a lot about Java, but also some programming principles.
+
+### Spring
+
+<img src = "/img/2021-05-27-TC-Guard/spring.png" class = "image left medium">
+
+The Spring framework is a Java platform that provides comprehensive infrastructure support for developing Java applications. 
+We started out by creating a small RESTful API using one of the tutorials on the official Spring website. 
+Alongside the tutorial we also used the extensive online documentation.
+
+In collage we had a .NET course where we used ASP.NET to build an API.
+We quickly discovered while working with Spring Boot that our preference goes to Spring instead of ASP.NET.
+One of the reasons is that Spring Boot does a lot for the developer while in ASP.NET you have to do more things manually.  
+
+### PostgreSQL
+The API's database is a postgreSQL database, which is used to store the collected data, the recorded trajectory controls and the drivers.
+Flyway is used to keep our database structure up-to-date using migrations.
+
+### API First
+To develop the API, we used the API-first design principle. This approach includes developing APIs that are consistent and reusable.
+This can be achieved by creating an API contract that describes the capabilities and behavior of the API.
+This ensures that more time is spent thinking about the design of the API.
+It often included additional planning and collaboration with our mentors.
+They provided feedback on the design of the API before any code was written.
+The first code that was written were tests.
+This method is called Test Driven Development (TDD).
+Specific attention was paid to the API contract, as the tests are based on them.
+TDD also allows for bugs to be found at an early stage because after writing code, the developers can run the tests and thus check whether the code works or not.
+
+# Android Application
+The user can add trajectory controls themselves via the record button.
+Once the user has passed the starting point of their trajectory control, they must press this button to start the process.
+During this process the starting point is saved and a checkpoint is added every 100 meters.
+When the user passes the end point of their trajectory control, they need to press this button again to define the end point and the maximum speed.
+The application will ask the user for the maximum speed, this will be automatically filled in by a speculation performed by the application.
```

Comment:
```suggestion
The application will ask the user for the maximum speed.
This will be automatically filled in by a speculation performed by the application.
```